src/rebar_compiler_dag.erl

```diff
@@ -100,13 +100,15 @@ safe_dir([H|T]) -> [H|safe_dir(T)].
 is_deleted_source(_G, _F, Extension, Extension, _ArtifactExt) ->
     %% source file
     true;
-is_deleted_source(_G, _F, Extension, _SrcExt, Extension) ->
-    %% artifact file - skip
-    false;
-is_deleted_source(G, F, _Extension, _SrcExt, _ArtifactExt) ->
-    %% must be header file or artifact
-    digraph:in_edges(G, F) == [] andalso maybe_rm_vertex(G, F),
-    false.
+is_deleted_source(G, F, Extension, _SrcExt, ArtifactExt) ->
+    case lists:member(Extension, ArtifactExt) of
+        true -> % artifact file, skip
+            false;
+        false ->
+            %% must be header file or artifact
+            digraph:in_edges(G, F) == [] andalso maybe_rm_vertex(G, F),
```

Comment:
Specifically here, the artifact file was possibly deleted more than required when the previous implementation didn't match on the extension (we pass in `[".beam"]` for `ArtifactExt`), but the .erl compiler used the on-disk build artifacts to re-fetch the compile opts as a fallback and likely hid the bug.

This might represent a perf gain?

Reply:
@max-au this is code you had initially written, I don't know if you still care about that aspect of the workflow.

Reply:
We didn't have any *.mib files or anything similar, so I guess I could never notice it. Change itself looks fine here, although it appears to be fragile enough to at least attempt a test...

Reply:
Yeah the test is implicitly covered by the MIB test. Without that change, the compiler permanently kept deleting build artifacts' from the DAG and recompiling them, which failed the test case. I'll look at making a more explicit one though.

Reply:
and done in https://github.com/erlang/rebar3/pull/2709/commits/661d79ceaaedd7fb33aa980d137036195b8c6efd

Reply:
That was a good call, turns out there were dangling artifact vertices in the DAG as well (the files were deleted and the edge removed when the source went away, but not the artifact vertex). The new regression test covers that as well.